source/common/grpc/common.cc

```diff
@@ -269,5 +286,98 @@ std::string Common::typeUrl(const std::string& qualified_name) {
   return typeUrlPrefix() + "/" + qualified_name;
 }
 
+struct BufferInstanceContainer {
+  BufferInstanceContainer(int ref_count, Buffer::InstancePtr buffer)
+      : ref_count_(ref_count), buffer_(std::move(buffer)) {}
+  std::atomic<int> ref_count_;
+  Buffer::InstancePtr buffer_;
+};
+
+static void derefBufferInstanceContainer(void* container_ptr) {
+  auto container = reinterpret_cast<BufferInstanceContainer*>(container_ptr);
+  container->ref_count_--;
+  if (container->ref_count_ <= 0) {
+    delete container;
+  }
+}
+
+grpc::ByteBuffer Common::makeByteBuffer(Buffer::InstancePtr bufferInstance) {
+  if (!bufferInstance) {
+    return {};
+  }
+  Buffer::RawSlice oneRawSlice;
+  // NB: we need to pass in >= 1 in order to get the real "n" (see Buffer::Instance for details).
+  int nSlices = bufferInstance->getRawSlices(&oneRawSlice, 1);
+  if (nSlices <= 0) {
+    return {};
+  }
+  auto container = new BufferInstanceContainer{nSlices, std::move(bufferInstance)};
+  if (nSlices == 1) {
+    grpc::Slice oneSlice(oneRawSlice.mem_, oneRawSlice.len_, &derefBufferInstanceContainer,
+                         container);
+    return {&oneSlice, 1};
+  }
+  STACK_ARRAY(manyRawSlices, Buffer::RawSlice, nSlices);
+  container->buffer_->getRawSlices(manyRawSlices.begin(), nSlices);
+  std::vector<grpc::Slice> slices;
+  slices.reserve(nSlices);
+  for (int i = 0; i < nSlices; i++) {
+    slices.emplace_back(manyRawSlices[i].mem_, manyRawSlices[i].len_, &derefBufferInstanceContainer,
+                        container);
+  }
+  return {&slices[0], slices.size()};
+}
+
+struct ByteBufferContainer {
+  ByteBufferContainer(int ref_count) : ref_count_(ref_count) {}
+  ~ByteBufferContainer() { ::free(fragments); }
+  std::atomic<int> ref_count_;
+  Buffer::BufferFragmentImpl* fragments = nullptr;
+  std::vector<grpc::Slice> slices_;
+};
+
+Buffer::InstancePtr Common::makeBufferInstance(const grpc::ByteBuffer& byteBuffer) {
+  auto buffer = std::make_unique<Buffer::OwnedImpl>();
+  if (byteBuffer.Length() == 0) {
+    return buffer;
+  }
+  // NB: ByteBuffer::Dump moves the data out of the ByteBuffer so we need to ensure that the
+  // lifetime of the Slice(s) exceeds our Buffer::Instance.
+  std::vector<grpc::Slice> slices;
+  byteBuffer.Dump(&slices);
+  if (slices.size() == 0) {
+    return buffer;
+  }
+  auto container = new ByteBufferContainer(static_cast<int>(slices.size()));
+  std::function<void(const void*, size_t, const Buffer::BufferFragmentImpl*)> releaser =
+      [container](const void*, size_t, const Buffer::BufferFragmentImpl*) {
+        container->ref_count_--;
+        if (container->ref_count_ <= 0) {
+          delete container;
+        }
+      };
+  // NB: addBufferFragment takes a pointer alias to the BufferFragmentImpl which is passed in so we
```

Comment:
Can we just make `fragments_` a `unique_ptr` and have it follow RAII?

Reply:
If I make it container->fragments_ = std::make_unique<Buffer::BufferFragmentImpl[]>(slices.size()); 
Then I get an error because BufferFragmentImpl is not copyable:

unique_ptr.h:831:34: error: no matching constructor for initialization of 'remove_extent_t<Envoy::Buffer::BufferFragmentImpl []> []' 

buffer_impl.h:430:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided

Could you suggest an alternative?

Reply:
Can `BufferFragmentImpl` be made copyable?

Reply:
We could, but it was explicitly made NonCopyable because it is aliased by OwnedImpl, and therefore copying it could break the pointer from OwnedImpl, so  I understand why it was made NonCopyable and I believe it needs to stay that way.

Any other ideas?

Reply:
In response to: "Can you elaborate?":

The BufferFragmentImpl is not owned by the OwnedImpl, instead it takes a reference to it, so allowing the BufferFragmentImpl to move could result in hard to find bugs if someone erroneously thought that implied that it was a value type as opposed to what it is: something with strong address-based identity.